### ISSUE TITLE
Change `cxg` conversion script to include `uns` data

### ISSUE THREAD

**Issue by kaloster** [chanzuckerberg/single-cell-data-portal#6659]
Right now, `uns` metadata that is needed for spatial embedding does not exist in `cxg`. Add the necessary code fro capturing this data and including it in `cxg` formatted file.

DoD:
- uns data is included and accessible in the `cxg` formatted dataset
- all tests that may break as a result should be fixed and pass

**Comment by tihuan** [comment]
@Bento007 @metakuni headsup that we might need to collaborate on updating the conversion script? Thank you!

**Comment by tihuan** [comment]
Also Jason from the Lattice Team said that they expect to get 5had updated to conform with schema 4.2, but we'll still need to convert them to cxg (manually?)

**Comment by metakuni** [comment]
> @Bento007 @metakuni headsup that we might need to collaborate on updating the conversion script? Thank you!

Thanks for the advanced ping, @tihuan. By 'collaborate', did you just mean some consultation and maybe PR review? The spatial work for Data Platform isn't scheduled to start until Mar 4, earliest.

Also note:
* The schema version for spatial work has been renamed from 4.2 to 5.1 (but timelines haven't changed)
* The [5.1 schema](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/5.1.0/schema.md) is still in draft, so may change.

**Comment by tihuan** [comment]
Thanks so much for all the context, @metakuni ! 

I think mostly I just want to make sure that we don't do double work 😆 

@kaloster has done some investigation on how to update the `cxg` converstion script to work with spatial data already, and we assumed that Data Platform probably will need to work on something similar when you guys tackle the spatial work stuff in March

So maybe we should just talk about how we should update the conversion script when Data Platform is ready to work on it!

**Comment by brianraymor** [comment]
@tihuan - Just FYI - Applications is responsible for the CXG validation and conversion per a decision that Arathi made prior to 4.0.

**Comment by tihuan** [comment]
Hey @brianraymor ! Ooh thanks for the headsup 💡 I see, so I should see who on the Applications team know more about validation and conversion, and go from there 😄 

In that case, @metakuni sorry for the false ping! 🙏

**Comment by metakuni** [comment]
> Hey @brianraymor ! Ooh thanks for the headsup 💡 I see, so I should see who on the Applications team know more about validation and conversion, and go from there 😄
> 
> In that case, @metakuni sorry for the false ping! 🙏

No worries, @tihuan !

As for whom on the Apps team... I believe that person was @joyceyan. 😏 For the schema 4.0 migration, she [worked on the Apps portion.](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/board?assignees=joyceyan&labels=4.0)

**Comment by metakuni** [comment]
One more thing worth noting: We'll have to coordinate when to merge and deploy your CXG conversion changes since that's part of the 5.1 schema, and we need to release all of those changes simultaneously. Hopefully, that won't complicate your development or testing schedules.

**Comment by tihuan** [comment]
Awesomee thanks for saving me time figuring out who I should reach out 😆 

Okay! Yeah coordination sounds good! I heard that Jason from the Lattice Team said that all the h5ad files should be updated to conform with 5.1 by Match 29, but I suppose we'd still need to deploy the new 5.1 schema conversion script to convert h5ad to cxg on Data Portal, and trigger the data processing pipeline to make sure everything is up to date?

**Comment by brianraymor** [comment]
>  I heard that Jason from the Lattice Team said that all the h5ad files should be updated to conform with 5.1 by Match 29

Yes - Jason was updating me on their readiness for re-curation for platform planning purposes.  The existing visium datasets will fail validation if they're not updated. 

> but I suppose we'd still need to deploy the new 5.1 schema conversion script to convert h5ad to cxg on Data Portal, and trigger the data processing pipeline to make sure everything is up to date?

Absolutely. When we migrate the data corpus to a new schema version, we need to re-create all the artifacts like seurat downloads and CXG(s) because the underlying data has changed (new/modified/deprecated metadata fields, replaced/obsolete ontology terms, new dataset citations, etc.)

**Comment by atarashansky** [comment]
@tihuan I'm intimately familiar with the CXG converter and also orchestrated a CXG remastering effort for the explorer performance improvements. Feel free to ping me as a consultant for this work.

**Comment by tihuan** [comment]
Oh you're a savior, @atarashansky 🤩🙌 Thanks so much! We'll definitely need your help with this lol

Can you help us with a quick rundown on the steps?

So far the pieces I know are:

1. Lattice Team will update all the h5ad files with the 5.1 schema

2. Apps Team needs to update the h5ad to cxg conversion script

And the goal is to update the data portal datasets with the latest schema, so we can enable Explorer spatial mode

I'm assuming the steps would be something like:

1. Apps Team update cxg conversion script, get it deployed to prod

2. Curators upload all the datasets that are now 5.1 schema

3. Manually trigger the processing pipeline to run the conversion script 

4. Once the pipeline is done, all the datasets will now have cxg format available

5. Profit 💰

Something like that? I don't know if 5.1 schema is backward compatible, if not, it feels like more coordination will be needed lol

CC: @kaloster @seve something we need to keep an eye on for rolling out the changes!

**Comment by tihuan** [comment]
Assigning this to me for now to drive the discussion with the schema team on potentially having them include uns data in the conversion script. Will re-assign back to Ronen, depending on the discusison outcome

**Comment by brianraymor** [comment]
RE

> Curators upload all the datasets that are now 5.1 schema
> 
> Manually trigger the processing pipeline to run the conversion script
> 

The migration of of the corpus from 5.0 -> 5.1 is automated. Curators do not download, revise, and upload datasets. There is no manual trigger for CXG and Seurat conversions. We can review the process on the DP-Apps call.

If you're curious, please see:

https://docs.google.com/document/d/1QNWicEb_C7W5B-L6Lb9jbnX897nUnoamEdaebznB_jw/edit#heading=h.jcf27gbiblyg
https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell/441
https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell/365

**Comment by tihuan** [comment]
Reached out to Dan about the collab possibility

**Comment by tihuan** [comment]
Putting this back to Disorder until DPT is ready for collab!

**Comment by tihuan** [comment]
I've updated the PR title and description given the meeting yesterday. Thanks so much again for the thorough presentation and info, @brianraymor !!

**Comment by tihuan** [comment]
=== OLD TICKET DESCRIPTION ===
Right now, `uns` metadata that is needed for spatial embedding does not exist in `cxg`. Add the necessary code fro capturing this data and including it in `cxg` formatted file.

DoD:
- uns data is included and accessible in the `cxg` formatted dataset
- all tests that may break as a result should be fixed and pass

**Comment by tihuan** [comment]
CC: @kaloster please help review the ticket and see if anything else is needed. Thanks so much!

**Comment by brianraymor** [comment]
Is there any additional processing/filtering required for these new `obs` fields? 

* array_row
* array_col

---

* in_tissue - already addressed by  https://github.com/chanzuckerberg/single-cell-explorer/issues/841 - sorry for the additional work. There was small group consensus to model this as an int instead of a bool because that's the source format and that's how it's modeled downstream by frameworks.

**Comment by tihuan** [comment]
Oh great question! I might need @kaloster to help answer this question 😆

**Comment by tihuan** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/6659?planning-poker) with Zenhub @kaloster @seve

**Comment by tihuan** [comment]
@kaloster @seve please help point this ticket thank you!

**Comment by tihuan** [comment]
@seve gentle ping thank you!

**Comment by tihuan** [comment]
@kaloster just missing your vote now 😊 Thanks!

**Comment by tihuan** [comment]
@kaloster should we close this ticket in favor of #839 https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell-explorer/839 ? Thank you!
